src/installer/corehost/cli/hostmisc/pal.unix.cpp

```diff
@@ -822,45 +823,11 @@ static void readdir(const pal::string_t& path, const pal::string_t& pattern, boo
             }
 
             // We are interested in files only
-            switch (entry->d_type)
+            int result = fstatat(dirfd(dir), entry->d_name, &sb, 0);
```

Comment:
This will call fstatat() for every entry, even if the system has d_type in struct dirent.  The reason some of these fields are populated in Linux is because fstat() can be very expensive -- especially if you're on a network file share; it's not a good idea to make it worse on all systems because of one system, especially since it's easy to support both.

One way to solve this:

```c
#if HAS_DIRENT_D_TYPE
int dirEntryType = entry->d_type;
#else
int dirEntryType = DT_UNKNOWN;
#endif

switch (dirEntryType)
{
   ...
}
``` 

Another way would be creating a static inline function that, given a `struct dirent` and a `DIR*`, returns the entry type (error handling omitted):

```
static inline get_dirent_type(const struct dirent *entry, const DIR *dir)
{
#if HAS_DIRENT_D_TYPE
    if (entry->d_type != DT_UNKNOWN && entry->d_type != DT_LNK) return entry;
#endif
    struct stat st;

    fstatat(dirfd(dir), entry->d_name, &st, 0);
    if (S_ISDIR(st.st_mode)) return DT_DIR;
    if (S_ISREG(st.st_mode)) return DT_REG;
    if (S_ISLNK(st.st_mode)) return DT_LNK;
    return DT_UNKNOWN;
}
```

Reply:
> it's not a good idea to make it worse on all systems because of one system

Are `DT_DIR` and `DT_REG` cases too common to qualify as hot path? If so, I agree.

Reply:
added the introspection and defined `DT_UNKNOWN`, `DT_DIR` macros for `!HAVE_DIRENT_D_TYPE` case.

Reply:
Previous commit was efbd452:
Assuming it is not a hot path (i.e. does not make a measurable-in-seconds difference for a average usecase), then I liked the previous approach as it was precise and cleaner, imo that is. 🙂

Reply:
The branches in the current version are removed by the compiler in the OSes where `d_type` isn't available, as it can do constant propagation and get rid of the switch statement altogether. :)

Reply:
I am not sure how branch elimination is related.

In the current version (master branch), based on the value of `d_type`, we either continue with the  loop, break the loop or calling `fstatat(2)`.

The PR's original change was to only call `fstatat(2)` in all cases, which is considered "bad for performance".

My argument about the required change (which we I have already made btw) was is it a hot (enough) path where perf matter which justifies code complexity by adding more #ifdef stew to the mixture? Most applications just do POSIX complaint implementations and there is also a value in one way of doing things.